Library/Homebrew/extend/object/deep_dup.rb

```diff
@@ -40,16 +43,35 @@ class Hash
   #
   #   hash[:a][:c] # => nil
   #   dup[:a][:c]  # => "c"
+  sig { returns(T.self_type) }
   def deep_dup
     hash = dup
     each_pair do |key, value|
-      if (::String === key && key.frozen?) || ::Symbol === key
-        hash[key] = value.deep_dup
+      case key
+      when ::String, ::Symbol
+        hash[key] = T.unsafe(value).deep_dup
       else
         hash.delete(key)
-        hash[key.deep_dup] = value.deep_dup
+        hash[T.unsafe(key).deep_dup] = T.unsafe(value).deep_dup
       end
     end
     hash
   end
 end
+
+class Module
```

Comment:
This is sourced from the current [upstream](https://github.com/rails/rails/pull/48106/files#diff-1219ce839b402dee52bc46245a0d30d13d961f2089e8cd306aedb978139792db) implementation